Code/Framework/AzToolsFramework/Tests/Viewport/ViewportEditorModeTests.cpp

```diff
@@ -0,0 +1,492 @@
+        EXPECT_EQ(m_viewportEditorModeTracker.GetViewportEditorModes({ viewportid }), nullptr);
+
+        // When a mode is deactivated for that viewport
+        const auto editorMode = ViewportEditorMode::Default;
+        UnitTest::ErrorHandler errorHandler(AZStd::string::format(
+            "Call to ExitMode for mode '%u' on id '%i' without precursor call to EnterMode", static_cast<AZ::u32>(editorMode), viewportid).c_str());
+        m_viewportEditorModeTracker.ExitMode({ viewportid }, editorMode);
+
+        // Expect a warning to be issued due to no precursor activation of that mode
+        EXPECT_EQ(errorHandler.GetExpectedWarningCount(), 1);
+
+        // Expect that viewport to now be tracked
+        const auto* viewportEditorModeState = m_viewportEditorModeTracker.GetViewportEditorModes({ viewportid });
+        EXPECT_TRUE(m_viewportEditorModeTracker.IsViewportModeTracked({ viewportid }));
+
+        // Expect the mode for that viewport to be inactive
+        EXPECT_NE(viewportEditorModeState, nullptr);
+        EXPECT_FALSE(viewportEditorModeState->IsModeActive(editorMode));
+    }
+
+    TEST_F(ViewportEditorModeTrackerTestFixture, GettingNonExistentViewportEditorModesForIdReturnsNull)
+    {
+        const ViewportId viewportid = 0;
+        EXPECT_FALSE(m_viewportEditorModeTracker.IsViewportModeTracked({ viewportid }));
+        EXPECT_EQ(m_viewportEditorModeTracker.GetViewportEditorModes({ viewportid }), nullptr);
+    }
+
+    TEST_F(ViewportEditorModeTrackerTestFixture, EnteringViewportEditorModesForExistingIdInThatStateIssuesWarningMsg)
+    {
+        // Given a viewport not currently tracked
+        const ViewportId viewportid = 0;
+        EXPECT_FALSE(m_viewportEditorModeTracker.IsViewportModeTracked({ viewportid }));
+        EXPECT_EQ(m_viewportEditorModeTracker.GetViewportEditorModes({ viewportid }), nullptr);
+
+        const auto editorMode = ViewportEditorMode::Default;
+        const auto expectedWarning = AZStd::string::format(
+            "Duplicate call to EnterMode for mode '%u' on id '%i'", static_cast<AZ::u32>(editorMode), viewportid);
+
+        {
+            UnitTest::ErrorHandler errorHandler(expectedWarning.c_str());
+
+            // When the mode is activated for the viewport
+            m_viewportEditorModeTracker.EnterMode({ viewportid }, editorMode);
+
+            // Expect no warning to be issued as there is no duplicate activation
+            EXPECT_EQ(errorHandler.GetExpectedWarningCount(), 0);
+
+            // Expect the mode to be active for the viewport
+            const auto* viewportEditorModeState = m_viewportEditorModeTracker.GetViewportEditorModes({ viewportid });
+            EXPECT_TRUE(m_viewportEditorModeTracker.IsViewportModeTracked({ viewportid }));
+            EXPECT_NE(viewportEditorModeState, nullptr);
+            EXPECT_TRUE(viewportEditorModeState->IsModeActive(editorMode));
+        }
```

Comment:
question: Could these two blocks be split into two separate tests? This does seem to be testing quite a lot for one test - may be necessary but just wanted to mention it (could be split in a later PR)

Reply:
Not really, as it does require that two-step approach to test the behavior. I do intend to do a refactor pass for this PR to encapsulate some of the boilerplate checks in functions so that should take the edge off of the bloat in this test 👍